compiler/ras/ILValidator.cpp

```diff
@@ -0,0 +1,356 @@
+ *
+ *      The Apache License v2.0 is available at
+ *      http://www.opensource.org/licenses/apache2.0.php
+ *
+ * Contributors:
+ *    Multiple authors (IBM Corp.) - initial implementation and documentation
+ ******************************************************************************/
+
+#include "ras/ILValidator.hpp"
+
+#include "il/DataTypes.hpp"
+#include "il/Node.hpp"
+#include "il/Node_inlines.hpp"
+#include "il/ILProps.hpp"
+#include "il/ILOps.hpp"
+#include "il/Block.hpp"
+#include "il/Block_inlines.hpp"
+#include "infra/Assert.hpp"
+
+#include <stdarg.h>
+
+#if defined(DEBUG) || defined(PROD_WITH_ASSUMES)
+#define ABORT() TR::trap()
+#else
+#define ABORT() comp()->failCompilation<TR::CompilationException>("Validation error")
+#endif
+
+#define FAIL() if (!feGetEnv("TR_continueAfterValidationError")) ABORT()
+
+TR::ILValidator::ILValidator(TR::Compilation *comp)
+   :_comp(comp)
+   ,_nodeStates(comp->trMemory())
+   ,_liveNodes(_nodeStates, comp->trMemory())
+   {
+   }
+
+TR::Compilation *TR::ILValidator::comp()
+   {
+   return _comp;
+   }
+
+void TR::ILValidator::checkSoundness(TR::TreeTop *start, TR::TreeTop *stop)
+   {
+   soundnessRule(start, start != NULL, "Start tree must exist");
+   soundnessRule(stop, !stop || stop->getNode() != NULL, "Stop tree must have a node");
+
+   TR::NodeChecklist treetopNodes(comp()), ancestorNodes(comp()), visitedNodes(comp());
+
+   // Can't use iterators here, because those presuppose the IL is sound.  Walk trees the old-fashioned way.
+   //
+   for (TR::TreeTop *currentTree = start; currentTree != stop; currentTree = currentTree->getNextTreeTop())
+      {
+      soundnessRule(currentTree, currentTree->getNode() != NULL, "Tree must have a node");
+      soundnessRule(currentTree, !treetopNodes.contains(currentTree->getNode()), "Treetop node n%dn encountered twice", currentTree->getNode()->getGlobalIndex());
+
+      treetopNodes.add(currentTree->getNode());
+
+      TR::TreeTop *next = currentTree->getNextTreeTop();
+      if (next)
+         {
+         soundnessRule(currentTree, next->getNode() != NULL, "Tree after n%dn must have a node", currentTree->getNode()->getGlobalIndex());
+         soundnessRule(currentTree, next->getPrevTreeTop() == currentTree, "Doubly-linked treetop list must be consistent: n%dn->n%dn<-n%dn", currentTree->getNode()->getGlobalIndex(), next->getNode()->getGlobalIndex(), next->getPrevTreeTop()->getNode()->getGlobalIndex());
+         }
+      else
+         {
+         soundnessRule(currentTree, stop == NULL, "Reached the end of the trees after n%dn without encountering the stop tree n%dn", currentTree->getNode()->getGlobalIndex(), stop? stop->getNode()->getGlobalIndex() : 0);
+         checkNodeSoundness(currentTree, currentTree->getNode(), ancestorNodes, visitedNodes);
+         }
+      }
+   }
+
+void TR::ILValidator::checkNodeSoundness(TR::TreeTop *location, TR::Node *node, NodeChecklist &ancestorNodes, NodeChecklist &visitedNodes)
```

Comment:
I didn't originally write this code, I only moved it from `compiler/ras/ILWalk.cpp`, so I can only make educated guesses for why it was written this way.

However, I believe the reason for calling the first parameter `location` is that it is only used to print the "location" of a failure in diagnostic messages; it is not actually used to check anything. In this sense, it keeps the interface consistent with the rest of the API.

I actually prefer the name `location` to `treetop` because it makes the purpose of the parameter clear: to specify where the node being checked is "located". However, I'm open to re-naming it to something like `treetopLocation` or `parentTreetop`. Both of these names make the relationship with the `node` parameter clear (at least in my head).